dot/parachain/candidate-validation/candidate_validation.go

```diff
@@ -155,82 +166,26 @@ func getValidationData(runtimeInstance parachainruntime.RuntimeInstance, paraID
-	buffer := bytes.NewBuffer(nil)
-	encoder := scale.NewEncoder(buffer)
-	err = encoder.Encode(pov)
-	if err != nil {
-		return nil, nil, false, fmt.Errorf("encoding pov: %w", err)
-	}
-	encodedPoVSize := buffer.Len()
-	if encodedPoVSize > int(persistedValidationData.MaxPovSize) {
-		return nil, nil, false, fmt.Errorf("%w, limit: %d, got: %d", ErrValidationInputOverLimit,
-			persistedValidationData.MaxPovSize, encodedPoVSize)
-	}
-
-	validationCodeHash, err := common.Blake2bHash([]byte(*validationCode))
-	if err != nil {
-		return nil, nil, false, fmt.Errorf("hashing validation code: %w", err)
-	}
-
-	if validationCodeHash != common.Hash(candidateReceipt.Descriptor.ValidationCodeHash) {
-		return nil, nil, false, fmt.Errorf("%w, expected: %s, got %s", ErrValidationCodeMismatch,
-			candidateReceipt.Descriptor.ValidationCodeHash, validationCodeHash)
-	}
+	*ValidationResult, error) {
 
-	// check candidate signature
-	err = candidateReceipt.Descriptor.CheckCollatorSignature()
+	persistedValidationData, validationCode, err := getValidationData(runtimeInstance,
+		candidateReceipt.Descriptor.ParaID)
 	if err != nil {
-		return nil, nil, false, fmt.Errorf("verifying collator signature: %w", err)
-	}
-
-	validationParams := parachainruntime.ValidationParameters{
-		ParentHeadData:         persistedValidationData.ParentHead,
-		BlockData:              pov.BlockData,
-		RelayParentNumber:      persistedValidationData.RelayParentNumber,
-		RelayParentStorageRoot: persistedValidationData.RelayParentStorageRoot,
+		return nil, fmt.Errorf("getting validation data: %w", err)
 	}
 
 	parachainRuntimeInstance, err := parachainruntime.SetupVM(*validationCode)
 	if err != nil {
-		return nil, nil, false, fmt.Errorf("setting up VM: %w", err)
-	}
-
-	validationResults, err := parachainRuntimeInstance.ValidateBlock(validationParams)
-	if err != nil {
-		return nil, nil, false, fmt.Errorf("executing validate_block: %w", err)
+		return nil, fmt.Errorf("setting up VM: %w", err)
 	}
 
-	candidateCommitments := parachaintypes.CandidateCommitments{
-		UpwardMessages:            validationResults.UpwardMessages,
-		HorizontalMessages:        validationResults.HorizontalMessages,
-		NewValidationCode:         validationResults.NewValidationCode,
-		HeadData:                  validationResults.HeadData,
-		ProcessedDownwardMessages: validationResults.ProcessedDownwardMessages,
-		HrmpWatermark:             validationResults.HrmpWatermark,
-	}
-
-	isValid, err := runtimeInstance.ParachainHostCheckValidationOutputs(
-		candidateReceipt.Descriptor.ParaID, candidateCommitments)
-	if err != nil {
-		return nil, nil, false, fmt.Errorf("executing validate_block: %w", err)
-	}
+	validationResults, err := validateFromExhaustive(parachainRuntimeInstance, *persistedValidationData,
+		*validationCode,
+		candidateReceipt, pov)
 
-	return &candidateCommitments, persistedValidationData, isValid, nil
+	return validationResults, err
```

Comment:
Let's check the error here, instead of returning result and error both.

```suggestion
	if err != nil {
		return nil, fmt.Errorf("validating from exhaustive: %w", err)
	}

	return validationResults, nil
```

Reply:
Updated.